### ISSUE TITLE
Plans for migrating to v0.10.0 for janestreet libraries?

### ISSUE THREAD

**Issue by mt-caret** [BinaryAnalysisPlatform/bap#786]
Hi,

I was just trying to build bap with the latest versions of Jane Street's libraries and got a bunch of errors; judging from https://github.com/BinaryAnalysisPlatform/bap/blob/master/opam/opam#L216-L217 I guess bap isn't currently compatible with building with, for example, core_kernel 0.10.0. Are there any plans to upgrading the packages that bap depends on in the near future?

**Comment by dbrumley** [comment]
I'm not the main author here, but one thing I've noticed is jane street seems to have not put a large priority on keeping APIs consistent throughout release.  LLVM does the same thing, and it sucks.

**Comment by ivg** [comment]
The 1.4 update brings compatibility with the Janestreet suite v0.9.x (and breaks compatibility with anything below `v0.9x`).  Most likely we will switch to `v0.10.x` soon, i.e., during the 1.5 release. Following Janestreet is not a goal in itself though, we are doing this mostly in reaction to the evolution of the OCaml infrastructure. Unfortunately, Janestreet treatment of API introduces lots of entropy in the OCaml world, that requires us to spend unnecessary effort to anticipate their ever changing opinions on how their abstractions should look like.

**Comment by ivg** [comment]
see also #756

**Comment by drvink** [comment]
I've been maintaining for a good two years now (just for my own personal use) a patchset that tracks the endless and infuriating Jane Street API churn. I thought about offering it here on multiple occasions but figured it would not be of any interest for exactly the reason @ivg mentions--unnecessary maintenance burden. However, I'm glad to make it available if anyone wants it.

**Comment by ivg** [comment]
@drvink, if you have an update to the v0.10.x then we will happily accept it. While 1.4 will still be dependent on v0.9.x, the 1.5 version is planned to be updated to the v0.10 version, and the sooner we will do this the better

**Comment by drvink** [comment]
Yes, I have patches against v0.10.116.08+75 and v0.10.116.16+14. If they don't already work fine on v0.10, it probably would not be much work to adapt them. (It was fortunate that I decided to check the issues list just now--I was just about to start updating them for v0.11.x!) I might be able to make a PR later, but if you would like to experiment with them yourself, I can just post them in a gist right now.

**Comment by ivg** [comment]
I would expect that v0.11 won't have as much breaking changes as v0.10 introduces on top of v0.9, and definitely won't be as breaking as v0.9 was to 113.x, though I might be wrong. 

We are now releasing once per 6 months, so if we will skip v0.10 and jump directly to v0.11, then that's fine. We can take both during this 6 months period, if it will make our life easier (less bumpy updates). Please, make the PR as soon as it is comfortable to you, there is no rush with this update, but since you've already done the majority of the work, there is no need to throw it away. The only limitation is that those patch should depend on versions that are released to the upstream repository.

**Comment by drvink** [comment]
There is one issue I'm not sure how to solve with regard to providing a "turn-key" patch: between v0.9 and v0.10, Jane Street removed the `ocamlbuild` plugin from `ppx_driver`. I dealt with this by simply pinning `ppx_driver` with a local patch that re-added the files they deleted (yes, the patch simply reverts the deletion--they _could_ have just kept it for compatibility's sake, but they leave no rug unpulled), but it doesn't seem so good to have to ask users to do something like this.

**Comment by ivg** [comment]
sheesh :( I believe, that the right solution would be to vendor the ocamlbuild plugin from ppx_driver into repository, however, if it doesn't have the MIT license, then we will just vendor it under the BinaryAnalysisPlatform umbrella.

**Comment by drvink** [comment]
OK, I've just completed a patch against OCaml 4.06.1 with `-safe-string` (as you no doubt know, this is the default as of 4.06), the latest (0.11.x) Jane Street libraries, and a few other things that also needed modification for `-safe-string`. It requires pinning of a number of dependencies.

Do you want this one as well? I'll prepare the patch against OCaml 4.05.x and the 0.10.x Jane Street libraries (it's basically already done, I just need to check it against the actual Core et al. 0.10 release) now, since the changes are much less invasive.

**Comment by ivg** [comment]
Both 0.10 and 0.11 drop the support of the 4.03.x version of OCaml, so both are equally impacting on the downstream users. The sooner we will start the update the more time they will have to update their systems. 

So far, it looks like that we can skip 0.10.x and jump over it directly to 0.11.x unless the latter have any conflicts with BAP dependencies.

**Comment by drvink** [comment]
I'll make the PR with the newer patch, then. It needs review: it compiles and tests pass, but I've been very conservative about use of unsafe features for this initial pass, which means that there is definitely more copying going on than is actually necessary. IIRC I've also changed the API in some places, but I'll leave it to your review and we can try to add or adjust things as necessary.

**Comment by XVilka** [comment]
Note, 0.12.x preview is available recently https://github.com/janestreet/base/commit/c8f6acfc1bff998f3630146e4f8c321182bc33bc

So it make sense to check against it too.
